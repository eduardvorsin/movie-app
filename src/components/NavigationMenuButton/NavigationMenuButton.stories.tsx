// eslint-disable-next-line import/no-extraneous-dependencies
import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/react';
import NavigationMenuButton, { Props } from './NavigationMenuButton';
import { MouseEventHandler, useState } from 'react';

const meta: Meta<typeof NavigationMenuButton> = {
	title: 'components/NavigationMenuButton',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the NavigationMenuButton'
		},
		testId: {
			description: 'Id for testing the component',
		},
		isActive: {
			description: 'Toggles the active state of the button',
		},
		onClick: {
			description: 'The callback that is called when clicked',
		}
	},
	component: NavigationMenuButton,
	parameters: {
		docs: {
			description: {
				component: 'The button that is used on mobile devices to open the main navigation of the site',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof NavigationMenuButton>;

const NavigationMenuButtonWithHooks = (props: Omit<Props, 'isActive' | 'onClick'>) => {
	const [isActive, setIsActive] = useState<boolean>(false);

	const clickHandler: MouseEventHandler<HTMLButtonElement> = (e) => {
		setIsActive((prevState) => !prevState);
		action('Clicked')(e);
	};

	return (
		<NavigationMenuButton
			{...props}
			isActive={isActive}
			onClick={clickHandler}
		/>
	);
}

export const Default: Story = {
	args: {
		className: '',
		testId: '',
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => <NavigationMenuButtonWithHooks {...args} />
};

export const Active: Story = {
	args: {
		isActive: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'The active state of the component',
			},
		},
	},
};

export const EventCallbacks: Story = {
	args: {
		className: '',
		testId: '',
	},
	parameters: {
		docs: {
			description: {
				story: 'When you click on the button, a callback is called',
			},
		},
	},
	render: (args) => <NavigationMenuButtonWithHooks {...args} />
};