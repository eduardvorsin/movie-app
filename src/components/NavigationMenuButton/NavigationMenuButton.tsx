'use client';

import { MouseEventHandler } from 'react';
import Button from '@/components/Button/Button';
import { GeneralProps } from '@/types/shared';

export type Props = {
	isActive: boolean,
	onClick: MouseEventHandler<HTMLButtonElement>,
	dictionary: Record<'active' | 'inactive', string>,
} & GeneralProps;

export default function NavigationMenuButton({
	isActive,
	onClick,
	className,
	testId,
	dictionary,
	...props
}: Props) {
	const topLineClasses = `origin-center transition-transform duration-150 ${isActive ? 'translate-y-[5.5px] translate-x-[5.8px] -rotate-45 scale-90' : ' translate-x-0 translate-y-0 rotate-0 scale-100'}`;

	const middleLineClasses = `origin-center transition-opacity duration-150 ${isActive ? 'opacity-0' : ' opacity-100'}`;

	const bottomLineClasses = `origin-center transition-transform duration-150 ${isActive ? '-translate-y-[5.5px] translate-x-[5.8px] rotate-45 scale-90' : ' translate-x-0 translate-y-0 rotate-0 scale-100'}`;

	return (
		<Button
			className={`relative z-100 ${className}`}
			appearance='secondary'
			iconButton
			onClick={onClick}
			testId={testId}
			{...props}
		>
			<svg
				className='p-2 w-[2.25rem] h-[2.25rem]'
				viewBox="0 0 32 32"
				fill="none"
			>
				<path
					className={topLineClasses}
					d="M2 7H30" stroke="currentColor" strokeWidth="3" strokeLinecap="round"
				/>
				<path
					className={middleLineClasses}
					d="M2 15.6667H30" stroke="currentColor" strokeWidth="3" strokeLinecap="round"
				/>
				<path
					className={bottomLineClasses}
					d="M2 24.3333H30" stroke="currentColor" strokeWidth="3" strokeLinecap="round" />
			</svg>

			{dictionary[isActive ? 'active' : 'inactive']}
		</Button>
	);
}