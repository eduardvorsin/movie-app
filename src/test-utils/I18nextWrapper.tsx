import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;