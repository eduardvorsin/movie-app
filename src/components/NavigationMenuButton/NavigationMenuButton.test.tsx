import { render, screen } from "@testing-library/react";
import NavigationMenuButton from "./NavigationMenuButton";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";
import userEvent from "@testing-library/user-event";

describe('NavigationMenuButton tests', () => {
	it('is rendered correctly', () => {
		const mockFn = jest.fn();
		const dictionary = {
			active: 'Close the navigation menu',
			inactive: 'Open the navigation menu'
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-navigation-menu-button')).toBeInTheDocument();
	});

	it('when clicking on a component, the mock function should be called', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		const dictionary = {
			active: 'Close the navigation menu',
			inactive: 'Open the navigation menu'
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>
		);

		await user.click(screen.getByRole<HTMLButtonElement>('button'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', () => {
		const mockFn = jest.fn();
		const dictionary = {
			active: 'Close the navigation menu',
			inactive: 'Open the navigation menu'
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-navigation-menu-button')).toMatchSnapshot();
	});

	it('is a snapshot when isActive is true', () => {
		const mockFn = jest.fn();
		const dictionary = {
			active: 'Close the navigation menu',
			inactive: 'Open the navigation menu'
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive
				testId='test-navigation-menu-button'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-navigation-menu-button')).toMatchSnapshot();
	});
});

describe('NavigationMenuButton integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('navigationMenuButton.inactive'));

		const mockFn = jest.fn();
		const dictionary = {
			active: i18next.t('navigationMenuButton.active'),
			inactive: i18next.t('navigationMenuButton.inactive')
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLButtonElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('navigationMenuButton.inactive'));

		const mockFn = jest.fn();
		const dictionary = {
			active: i18next.t('navigationMenuButton.active'),
			inactive: i18next.t('navigationMenuButton.inactive')
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLButtonElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		const mockFn = jest.fn();
		const dictionary = {
			active: i18next.t('navigationMenuButton.active'),
			inactive: i18next.t('navigationMenuButton.inactive')
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-navigation-menu-button')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		const mockFn = jest.fn();
		const dictionary = {
			active: i18next.t('navigationMenuButton.active'),
			inactive: i18next.t('navigationMenuButton.inactive')
		};
		render(
			<NavigationMenuButton
				onClick={mockFn}
				dictionary={dictionary}
				isActive={false}
				testId='test-navigation-menu-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-navigation-menu-button')).toMatchSnapshot();
	});
});