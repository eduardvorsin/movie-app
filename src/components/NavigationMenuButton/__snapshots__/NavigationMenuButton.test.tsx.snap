// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationMenuButton integration tests is a snapshot with English localization 1`] = `
<button
  class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   relative z-100 undefined"
  data-testid="test-navigation-menu-button"
>
  <span
    class="inline-flex items-center"
  >
    <svg
      class="p-2 w-[2.25rem] h-[2.25rem]"
      fill="none"
      viewBox="0 0 32 32"
    >
      <path
        class="origin-center transition-transform duration-150  translate-x-0 translate-y-0 rotate-0 scale-100"
        d="M2 7H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-opacity duration-150  opacity-100"
        d="M2 15.6667H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-transform duration-150  translate-x-0 translate-y-0 rotate-0 scale-100"
        d="M2 24.3333H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
    </svg>
    Open the navigation menu
  </span>
</button>
`;

exports[`NavigationMenuButton integration tests is a snapshot with Russian localization 1`] = `
<button
  class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   relative z-100 undefined"
  data-testid="test-navigation-menu-button"
>
  <span
    class="inline-flex items-center"
  >
    <svg
      class="p-2 w-[2.25rem] h-[2.25rem]"
      fill="none"
      viewBox="0 0 32 32"
    >
      <path
        class="origin-center transition-transform duration-150  translate-x-0 translate-y-0 rotate-0 scale-100"
        d="M2 7H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-opacity duration-150  opacity-100"
        d="M2 15.6667H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-transform duration-150  translate-x-0 translate-y-0 rotate-0 scale-100"
        d="M2 24.3333H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
    </svg>
    Открыть меню навигации
  </span>
</button>
`;

exports[`NavigationMenuButton tests is a basic snapshot 1`] = `
<button
  class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   relative z-100 undefined"
  data-testid="test-navigation-menu-button"
>
  <span
    class="inline-flex items-center"
  >
    <svg
      class="p-2 w-[2.25rem] h-[2.25rem]"
      fill="none"
      viewBox="0 0 32 32"
    >
      <path
        class="origin-center transition-transform duration-150  translate-x-0 translate-y-0 rotate-0 scale-100"
        d="M2 7H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-opacity duration-150  opacity-100"
        d="M2 15.6667H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-transform duration-150  translate-x-0 translate-y-0 rotate-0 scale-100"
        d="M2 24.3333H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
    </svg>
    Open the navigation menu
  </span>
</button>
`;

exports[`NavigationMenuButton tests is a snapshot when isActive is true 1`] = `
<button
  class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   relative z-100 undefined"
  data-testid="test-navigation-menu-button"
>
  <span
    class="inline-flex items-center"
  >
    <svg
      class="p-2 w-[2.25rem] h-[2.25rem]"
      fill="none"
      viewBox="0 0 32 32"
    >
      <path
        class="origin-center transition-transform duration-150 translate-y-[5.5px] translate-x-[5.8px] -rotate-45 scale-90"
        d="M2 7H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-opacity duration-150 opacity-0"
        d="M2 15.6667H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
      <path
        class="origin-center transition-transform duration-150 -translate-y-[5.5px] translate-x-[5.8px] rotate-45 scale-90"
        d="M2 24.3333H30"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="3"
      />
    </svg>
    Close the navigation menu
  </span>
</button>
`;
